docs/source/developer_guide/library_design.md

```diff
@@ -0,0 +1,111 @@
+Read [geoarrow format documentation](https://github.com/geopandas/geo-arrow-spec/blob/main/format.md) specification for more detail.
+
+Cuspatial implements partial arrow union array via `GeoColumn` and `GeoMeta`.
+A GeoColumn is a composition of child columns and a `GeoMeta` object.
+The `GeoMeta` owns two arrays that are similar to the types buffer and offsets buffer from dense union.
+
+```{note} Currently,
+`GeoColumn` only implements four concrete array types: `points`, `multipoints`, `multilinestrings` (named as lines) and `multipolygon` (named as polygons).
+Linestrings and multilinestrings are stored uniformly as multilinestrings in the multilinestrings array.
+Polygons and multipolygons are stored uniformly as multipolygons in multipolygon array.
+Points and multipoints are stored separately in different arrays.
+This is because storing points in a multipoints array requires 50% more storage overhead.
+While this may also be true for linestrings and polygons,
+previous use cases of cuspatial often involves more complex linestrings and polygons,
+and the storage overhead is lower compared to points.
+```
+
+`GeoSeries` and `GeoDataFrame` inherits from `cudf.Series` and `cudf.DataFrame` respectively.
+`Series` and `DataFrame` are all generic `Frame` objects which represents a collection of generic columns.
+cuSpatial extends these cudf objects by allowing `GeoColumn` to present in the frame.
+
+`GeoSeries` and `GeoDataFrame` are convertible to and from `geopandas`.
+Interoperability between cuspatial, `geopandas` and other data format is maintained in `cuspatial.io` package.
+
+### UnionArray Compliance
+
+As previously mentioned,
+cuspatial's `GeoColumn` is an specialization of arrow's `UnionArray`.
+A fundamental addition in data types should be implemented in `cudf` and `GeoColumn` should simply inherits its functionality.
+However,
+`UnionArray` stands distinctly from existing data types in libcudf and requires substantial effort to implement.
+In the interim,
+cuspatial developers should build `GeoColumn` complying to `UnionArray` standards.
+Such effort may be upstreamed to `cudf` when development is more underway.
+
+## GIS computation APIs
+
+Besides the data structure,
+cuspatial maintains a set of computation APIs.
+The computation APIs are organized into several modules.
+All spatial computation modules are further grouped into a `spatial` subpackage.
+Developers are encouraged to use specific names for the module of the function added in the PR.
+
+### Legacy and Modern APIs
+
+For historical reasons,
+cuspatial APIs exposes raw array inputs for users to provide raw geometry coordinate array and offsets.
+Newer APIs should accept a `GeoSeries` or `GeoDataFrame` as input.
+Developers may extract geometry offsets and coordinates via geometry accessors provided,
+such as `GeoSeries.point`, `GeoSeries.multipoint`, `GeoSeries.lines`, `GeoSeries.polygon`.
```

Comment:
Why is `GeoSeries.lines` plural while all the others are singular?

```suggestion
### Legacy and Modern APIs

For historical reasons, older cuSpatial APIs expose raw array inputs for
users to provide raw geometry coordinate arrays and offsets. Newer Python
APIs should accept a `GeoSeries` or `GeoDataFrame` as input. Developers
may extract geometry offsets and coordinates via cuSpatial's geometry
accessors such as `GeoSeries.point`, `GeoSeries.multipoint`,
`GeoSeries.lines`, `GeoSeries.polygon`.
```

This point about accessors seems to have nothing to do with legacy vs. modern. Perhaps that should be a separate section?

Reply:
In fact the accessors are all in plural form:

```suggestion
### Legacy and Modern APIs

For historical reasons, older cuSpatial APIs expose raw array inputs for
users to provide raw geometry coordinate arrays and offsets. Newer Python
APIs should accept a `GeoSeries` or `GeoDataFrame` as input. Developers
may extract geometry offsets and coordinates via cuSpatial's geometry
accessors such as `GeoSeries.points`, `GeoSeries.multipoints`,
`GeoSeries.lines`, `GeoSeries.polygons`.
```

As mentioned in one of the comment above, `lines` should be renamed to `multilinestrings`, `polygons` should be renamed to `multipolygons`.